Scalar wave executable

nilsdeppe opened the pull request:
## Proposed changes

Note: Depends on #493 (which depends on #491). This means only the commits "Add tests for input files", "Add Parallel Compenent DgElementArray", and "Add 1D, 2D and 3D ScalarWave executables" are new.

- Close #505
- Adds testing infrastructure for input files. This should be flexible enough to allow easy extension of it in the future. Currently all input files are tested for parsing, this is not optional. It is possible to opt-in to having the executable run. Output is not checked, though.
- Add what is expected to be the core DgElementArray parallel component. I expect this to grow a bit as more features are added to the code, but this is enough for what we need right now.
- Add ScalarWave executables in 1D, 2D and 3D.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`


### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
-->

nilsdeppe committed:
Add tests for input files

nilsdeppe committed:
Add Parallel Component DgElementArray

The plan is that this Parallel component will be the component used
for the dG elements in all dG evolutions. I anticipate that it will
change, especially as the dg::Actions::InitializeElement action
changes. However, this provides the core functionality needed to do an
evolution.

nilsdeppe committed:
Add 1D, 2D, and 3D ScalarWave executables

Add the executables, as well as input files for them. The input files
are tested using the infrastructure added previously.

nilsdeppe committed:
Fix Main.hpp to work on GCC

osheamonn commented on src/Evolution/DiscontinuousGalerkin/DgElementArray.hpp:
```diff
@@ -0,0 +1,91 @@
+template <class Metavariables, class ActionList>
+void DgElementArray<Metavariables, ActionList>::initialize(
+    Parallel::CProxy_ConstGlobalCache<Metavariables>& global_cache,
+    std::unique_ptr<DomainCreator<volume_dim, Frame::Inertial>> domain_creator,
```
Can this be a const ref? Does it even matter for a `unique_ptr`?

osheamonn commented on src/Evolution/DiscontinuousGalerkin/DgElementArray.hpp:
```diff
@@ -0,0 +1,91 @@
+void DgElementArray<Metavariables, ActionList>::initialize(
+    Parallel::CProxy_ConstGlobalCache<Metavariables>& global_cache,
+    std::unique_ptr<DomainCreator<volume_dim, Frame::Inertial>> domain_creator,
+    double initial_time, double initial_dt) noexcept {
```
Should both these be const?

osheamonn commented on src/Evolution/Systems/ScalarWave/System.hpp:
```diff
@@ -1,4 +1,3 @@
-
```
Why the extra newline?

nilsdeppe replied:
I removed the extra newline. Is that an issue?

osheamonn replied:
I must've been tired and thought the newline was being added...

osheamonn commented on tests/InputFiles/ScalarWave/Input1DPeriodic.yaml:
```diff
@@ -0,0 +1,30 @@
+  AdamsBashforthN:
+    TargetOrder: 3
+
+NumericalFlux:
```
Why does this work? I would've thought no numerical flux being specified would be an error.

kidder replied:
NumericalFlux -> NumericalFluxParameters

osheamonn commented on tests/InputFiles/ScalarWave/Input1DPeriodic.yaml:
```diff
@@ -0,0 +1,30 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+# Executable: EvolveScalarWave1D
```
I don't feel particularly strongly about this, but I like the idea of there being a newline after the licence. I find it makes it easier to parse since I've gotten into the habit of ignoring the licence text.

kidder commented on cmake/AddInputFileTests.cmake:
```diff
@@ -0,0 +1,130 @@
+
+function(add_single_input_file_test INPUT_FILE EXECUTABLE CHECK_TYPE TIMEOUT)
+  # Extract just the name of the input file
+  string(FIND "${INPUT_FILE}" "/" POSITION_OF_LAST_SLASH REVERSE)
```
get_filename_component

kidder commented on cmake/AddInputFileTests.cmake:
```diff
@@ -0,0 +1,130 @@
+  string(FIND "${INPUT_FILE}" "tests/InputFiles/" POSITION_OF_INPUT_FILE_DIR)
+  math(EXPR
+    POSITION_OF_INPUT_FILE_DIR
+    ${POSITION_OF_INPUT_FILE_DIR}+17
```
comment magic number

kidder commented on cmake/AddInputFileTests.cmake:
```diff
@@ -0,0 +1,130 @@
+    )
+
+  string(SUBSTRING "${INPUT_FILE}" ${POSITION_OF_INPUT_FILE_DIR}
+    ${LENGTH_OF_EXECUTABLE_CLASS}
```
EXECUTABLE_CLASS -> EXECUTABLE_DIR_NAME

kidder commented on cmake/AddInputFileTests.cmake:
```diff
@@ -0,0 +1,130 @@
+# test for each one. The input files must contain a line of the form:
+# '# Executable: EvolveScalarWave1D'
+# with the name of the executable that should be able to parse and run
+# using the input file, and a line one the form:
```
of the form

kidder commented on cmake/AddInputFileTests.cmake:
```diff
@@ -0,0 +1,130 @@
+    # "parse" is ignored because it's always run, empty is accepted.
+    string(REGEX MATCH "#[ ]*Check:[^\n]+"
+      INPUT_FILE_CHECKS "${INPUT_FILE_CONTENTS}")
+    # Extract executable name, and remove trailing white space
```
please

kidder commented on cmake/AddInputFileTests.cmake:
```diff
@@ -0,0 +1,130 @@
+        "parse"
+        ${TIMEOUT}
+        )
+    foreach(CHECK_TYPE ${INPUT_FILE_CHECKS})
```
list removing duplicates

nilsdeppe committed:
Allow conversion ElementId -> ArrayIndex<ElementIndex>

nilsdeppe committed:
Improve termination message from std::terminate

nilsdeppe committed:
Add missing tmpl::pin to ConstGlobalCache

nilsdeppe committed:
Add NumericalFluxParameters global cache tag

kidder commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Tags.hpp:
```diff
@@ -0,0 +1,19 @@
+ */
+template <typename NumericalFluxType>
+struct NumericalFluxParams {
+  static constexpr OptionString help = "The type of the numerical flux";
```
The options for the numerical flux

kidder commented on src/Evolution/DiscontinuousGalerkin/DgElementArray.hpp:
```diff
@@ -0,0 +1,92 @@
+
+template <class Metavariables, class ActionList>
+struct DgElementArray {
+  static constexpr const size_t volume_dim = Metavariables::system::volume_dim;
```
no const

kidder commented on src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp:
```diff
@@ -0,0 +1,78 @@
+                 Actions::UpdateU, Actions::AdvanceTime, Actions::FinalTime>>>;
+
+  static constexpr const char* const help{
+      "Evolve a Scalar Wave in 1 spatial dimension"};
```
1 -> dim

kidder replied:
describe metavariables

kidder commented on src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp:
```diff
@@ -0,0 +1,78 @@
+                 dg::Actions::ComputeBoundaryFlux<EvolutionMetavars>,
+                 Actions::UpdateU, Actions::AdvanceTime, Actions::FinalTime>>>;
+
+  static constexpr const char* const help{
```
OptionString

kidder commented on src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp:
```diff
@@ -0,0 +1,78 @@
+
+  static constexpr const char* const help{
+      "Evolve a Scalar Wave in 1 spatial dimension"};
+  static constexpr bool ignore_unrecognized_command_line_options = false;
```
remove

kidder commented on src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp:
```diff
@@ -0,0 +1,78 @@
+
+  enum class Phase {
+    Initialization,
+    PerformSingletonAlgorithm,
```
remove

nilsdeppe committed:
Properly reset Variables tags after mutate

Fix #505

nilsdeppe commented:
Okay, I've updated the PR addressing the comments and adding (as the new first commit) the fix to DataBox that is in #505, which is needed for SW to work
